### PR TITLE
scroll-while-dragging

### PR DESCRIPTION
**Pull request by zeeamber**
fix for Scroll while dragging #121

### CODE REVIEW

**Comment by bevacqua** [comment]
Hey there, this seems like a much leaner solution: https://github.com/luckylooke/dragular/commit/0871a581613555ad94da54b8daed631a254903a7, he uses the `wheel` event to propagate the scroll amount using `e.deltaY`. While we're on it, consider updating `el.scrollLeft` as well, using `e.deltaX`. Mind taking a look?

+@luckylooke

**Comment by zeeamber** [comment]
sure, I'll try to update it for scrollLeft as well. I'm trying not to use any other library like jquery, angular, etc

**Comment by bevacqua** [comment]
Oh no I just meant looking at that changeset to see how he handles it with `wheel` events, as it seems much cleaner (and lots less of code, too!)

**Comment by zeeamber** [comment]
Oh got it. Yeah sure I will check that

**Comment by bevacqua** [comment]
Cool! :)

**Comment by bevacqua** [comment]
Hey @zeeamber do you have a chance to look at fixing the code?

**Comment by zeeamber** [comment]
Hi @bevacqua, I tried to implement  using wheel event but sadly no luck. can you give me some time to optimize it other way around if find some. I think i should be able to reduce some LOC

**Comment by zeeamber** [comment]
Hi @bevacqua, both horizontal and vertical scrolls are working now. I tested it for chrome, firefox, & safari.

**Comment by bevacqua** [comment]
What about mobile?

**Comment by zeeamber** [comment]
@bevacqua, I just tested it with mobile. its working fine.

File: dist/dragula.js [hunk 1]
```diff
@@ -327,6 +342,66 @@ function dragula (initialContainers, options) {
...
+    
+  function stopScroll() {
+      o.scrolling = false;
+      clearTimeout(_scrollTimer);
+  }
+    
+  function scroll() {
+      updateWinScroll();
+      var dropTarget = _item.parentElement;
+      var renderMode = navigator.userAgent.toLowerCase().indexOf('firefox') !== -1 ? documentElement : body ;
```
Comment (bevacqua) [inline dist/dragula.js]: We should favor feature detection

File: dist/dragula.js [hunk 2]
```diff
@@ -37,6 +42,7 @@ function dragula (initialContainers, options) {
   if (o.direction === void 0) { o.direction = 'vertical'; }
   if (o.delay === void 0) { o.delay = false; }
   if (o.delay === true) { o.delay = 300; }
+  if (o.scrolling === void 0) { o.scrolling = false; }
```
Comment (bevacqua) [inline dist/dragula.js]: Shouldn't be an option

**Comment by bevacqua** [comment]
The demo isn't working for me on Chrome

We should try and implement this using `wheel` events [_(like here)_](https://github.com/luckylooke/dragular/commit/0871a581613555ad94da54b8daed631a254903a7), seems much less involved too

**Comment by zeeamber** [comment]
@bevacqua  I'm still on to implement it using wheel events. I know using wheel event is the best solution here. hope I succeed soon

**Comment by bevacqua** [comment]
Oh cool

**Comment by luckylooke** [comment]
I have made different solution since touch-devices don't have wheels, see [demo] (http://luckylooke.github.io/dragular/#/docs/exampleScrollingDrag) and [changes in code](https://github.com/luckylooke/dragular/commit/05f47849544b090cdfa949995c088c40046ca4c5#diff-8bd31747c514dcad2e281d0ddaaee0ae). I have still left wheel listener for devices with wheel:
```js
// wheel event listener
function scrollContainer(e) {
      if (_targetContainer) {
        var before = _targetContainer.scrollTop;
        _targetContainer.scrollTop += e.deltaY;
        // block scroll of the document when container can be scrolled
        if (before !== _targetContainer.scrollTop) {
          e.stopPropagation();
          e.preventDefault();
        }
      }
    }
```

**Comment by bevacqua** [comment]
Do these solutions work when the user is dragging something (but not moving it, just sitting) over the edge of a container? Kind of like how you can scroll a Finder window while dragging something just by hovering over the bottom of the window?

**Comment by luckylooke** [comment]
@bevacqua yes, there is interval used, but its not part of library, you need just simple event listeners to make it work.

**Comment by luanfonceca** [comment]
I just grabbed the code and run it, i didnt change anything on the dragula.js or in the configurations, but i am getting a weird behaviour, i just clicked and the page scrolled up: 



For me, this feature of scroll the page needs to work just when the item in on the borders, like if some piece of the `div` touched any side of the page.


o/

**Comment by zeeamber** [comment]
Hi @luanfonceca. thanks for raising the bug. did some changes in scroll criteria. should work properly now.

**Comment by bevacqua** [comment]
Agree that functionality should trigger just near edges, not during scroll

**Comment by bevacqua** [comment]
Also, no browser sniffing. Figure out another condition

**Comment by luanfonceca** [comment]
Hi @zeeamber, i just tested that behaviour and it is ok, but i found another one: 
When i drag the item to outside of his container, the scroll doesn't work anymore:

**Comment by bevacqua** [comment]
That's what I would expect

**Comment by luanfonceca** [comment]
Ok, then everything is OK for me

o/

Em sáb, 22 de ago de 2015 7:55 AM, Nicolas Bevacqua <
notifications@github.com> escreveu:

**Comment by bevacqua** [comment]
@zeeamber could you get rid of browser sniffing and merge with master? I can take it from there

**Comment by zeeamber** [comment]
yeah sure I will give it a try. 
@bevacqua

**Comment by zeeamber** [comment]
merged with master @bevacqua

**Comment by bevacqua** [comment]
Doesn't look that way

**Comment by luanfonceca** [comment]
Hi @zeeamber, sorry for ask this, but did you how to update your code with the origin one, I mean, pull the the code from this repo, merge it with your own fork and fixing the conflicts, all that?!

o/

**Comment by zeeamber** [comment]
Hey @luanfonceca , yeah little confusion working on git repo, I'm leafing through the documentation though :)

**Comment by zeeamber** [comment]
I pulled the changes from orig master, updated it and merged with my master branch. 
can you check it now if its okay? @luanfonceca @bevacqua

**Comment by luanfonceca** [comment]
Hi folks, i think i dont have so much knowledge about what code was in the master, but it seems everything is fine for me.

**Comment by lettfeti** [comment]
Aren't the conflicts resolved? Is there something else stopping this PR from being merged?

**Comment by zeeamber** [comment]
Conflicts are resolved. I think @bevacqua can take it from here

**Comment by btecu** [comment]
Looking forward this feature.

**Comment by bevacqua** [comment]
Can't merge this yet

**Comment by zeeamber** [comment]
All checks have passed here. can u merge it now ? @bevacqua @btecu

**Comment by bevacqua** [comment]
Yeah, please squash the commits. Thanks

**Comment by btecu** [comment]
I think the code should also go in `dragula.js` not just `dist/dragula.js`.

**Comment by bevacqua** [comment]
Agreed

**Comment by mfsousa** [comment]
@bevacqua about what @luanfonceca told when dragging outside of container. I don't agree: I expect to continue scrolling. I say more: even when we drag some item outside of window (window is resized and we move cursor, dragging item, until we have the cursor outside of the window) it should continue scrolling... and not only scroll when we have the cursor inside the container.

And also, we should take in consideration the drag distance to set the scroll speed: take in consideration the starting dragging point and the current point we are dragging and set the scroll speed according to that...

Example > do like trello is doing: https://trello.com/b/Ragttcss/drakula (make a trello board like this one and see how trello does it) ;)
You see that they don't limit the scroll by just the container where we drag

**Comment by bevacqua** [comment]
+1 drag speed

**Comment by swatigoel1991** [comment]
i took the code and run it, it only works in chrome for me. It doesn't work on Mozilla and IE9 browser. I din't check IE9+ browsers though.

**Comment by bevacqua** [comment]
Closing in favor of modules such as [`dom-autoscroller`](https://www.npmjs.com/package/dom-autoscroller)

**Comment by nguyenj** [comment]
@zeeamber I have a [PR](https://github.com/bevacqua/dragula/pull/449) that address scrolling while dragging. Let me know if this works for you.

**Pull request closed without merging**